pre-docs/writing_yaml_plans.md

```diff
@@ -473,3 +473,134 @@ plan mymodule::yamlplan(
+}
+```
+If the author were writing the code as a Puppet plan they would likely not employ the use of the lambda. In this example the converted Puppet code is 'correct' but is not particularly natural or readable. 
+
+### Resource step variable interpolation
+
+When applying puppet resources in a resource step variable interpolation will behave differently between YAML and Puppet plans. In order to illustrate an important distinction consider the following yaml plan:
+```
+steps:
+  - description: apply prep
+    eval: >
+      apply_prep('localhost')
+  - target: localhost
+    description: Apply a file resource
+    resources:
+    - type: file
+      title: '/tmp/foo'
+      parameters:
+        content: $facts['os']['family']
+        ensure: present
+  - name: file_contents
+    description: Read contents of file managed with file resource
+    eval: >
+      file::read('/tmp/foo')
+      
+return: $file_contents
+```
+This plan performs apply prep on a localhost target, next it uses a Puppet `file` resource to write the OS family discovered from the Puppet `$facts` hash to a temporary file. Finally it reads the value written to the file and returns it. Running `bolt plan convert` on this plan produces the following puppet code:
+```
+plan yaml_plans::interpolation_pp() {
+  apply_prep('localhost')
+
+  $interpolation = apply('localhost') {
+    file { '/tmp/foo':
+      content => $facts['os']['family'],
+      ensure => 'present',
+    }
+  }
+  $file_contents = file::read('/tmp/foo')
+
+  return $file_contents
+}
+```
+In this case the Puppet plan produced from the Yaml plan works as expected where the Yaml version fails. The cause of the failure is that the `$facts` variable is attempted to be resolved as a plan variable instead of being evaluated as part of compiling the manifest code in an apply block.
+
+The next example also has a somewhat related interpolation issue. 
+
+## Dependency order
+
+The resources in a resource list are applied in order. It is still possible to explicitly set dependencies but it is important to note how to reference them. Consider the following Yaml plan:
+```
+parameters:
+  nodes:
+    type: TargetSpec
+steps:
+  - eval: |
+      apply_prep($nodes)
+  - name: pkg
+    target: $nodes
+    resources:
+      - title: openssh-server
+        type: package
+        parameters:
+          ensure: present
+          before: File['/etc/ssh/sshd_config']
+      - title: /etc/ssh/sshd_config
+        type: file
+        parameters:
+          ensure: file
+          mode: '0600'
+          content: ''
+          require: Package['openssh-server']
+```
+Executing this plan will fail during catalog compilation because of how the resources referenced in the `before` and `require` parameters are parsed. Specifically the error will be at `Could not find resource 'File['/etc/ssh/sshd_config']' in parameter 'before'`. The solution for this is to not quote the resource titles. See the corrected plan example below.
```

Comment:
Recommend changing `see the corrected plan example below.` to `Here is the corrected plan example:` This is because when I first read this I wasn't sure whether "below" meant way farther down in the document, or immediately below.